File: scripts/run_rag_eval.py
#!/usr/bin/env python3
"""
Exécute le jeu de questions RAG contre le webhook n8n et écrit un CSV de résultats.

Variables d'environnement (optionnel) :
  RAG_EVAL_WEBHOOK_URL   URL du POST (défaut : http://127.0.0.1:5678/webhook/chat)
  RAG_EVAL_DELAY_SEC     Pause entre requêtes (défaut : 0.5) — limite charge / rate limits
  RAG_EVAL_TIMEOUT_SEC   Timeout HTTP (défaut : 180)

Réglages n8n (côté serveur n8n / Docker) pour ajuster retrieval sans rééditer le JSON :
  RAG_PREFETCH_LIMIT       candidats prefetch hybride (défaut 24)
  RAG_FINAL_LIMIT          points après fusion RRF (défaut 10)
  RAG_RERANK_TOP_N         top_n Cohere rerank (défaut 6)
  RAG_MAX_CONTEXT_CHARS    plafond caractères injectés au LLM (défaut 2400)
  RAG_LLM_MAX_CHUNKS       nombre max d'extraits dans le prompt (défaut 8)
  COHERE_RERANK_ENABLED    false | 0 | off pour désactiver Cohere (passe-through hybride)
"""

from __future__ import annotations

import argparse
import csv
import os
import sys
import time
import urllib.error
import urllib.request
from datetime import datetime, timezone


def post_json(url: str, body: bytes, timeout: float) -> tuple[int, str]:
    req = urllib.request.Request(
        url,
        data=body,
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            return resp.getcode(), resp.read().decode("utf-8", errors="replace")
    except urllib.error.HTTPError as e:
        return e.code, (e.read() or b"").decode("utf-8", errors="replace")
    except urllib.error.URLError as e:
        return -1, str(e.reason if hasattr(e, "reason") else e)


def main() -> int:
    p = argparse.ArgumentParser(description="Run RAG eval CSV against n8n webhook")
    p.add_argument(
        "--csv",
        default=os.path.join(os.path.dirname(__file__), "..", "tests", "rag_eval_questions.csv"),
        help="Fichier CSV d'entrée (colonnes id, categorie, question, ...)",
    )
    p.add_argument(
        "--out",
        default="",
        help="CSV de sortie (défaut: tests/rag_eval_results_<timestamp>.csv)",
    )
    p.add_argument(
        "--url",
        default=os.environ.get("RAG_EVAL_WEBHOOK_URL", "http://127.0.0.1:5678/webhook/chat"),
        help="URL webhook chat",
    )
    p.add_argument(
        "--delay",
        type=float,
        default=float(os.environ.get("RAG_EVAL_DELAY_SEC", "0.5")),
        help="Secondes entre chaque requête",
    )
    p.add_argument(
        "--timeout",
        type=float,
        default=float(os.environ.get("RAG_EVAL_TIMEOUT_SEC", "180")),
        help="Timeout HTTP par question",
    )
    args = p.parse_args()

    in_path = os.path.abspath(args.csv)
    if not os.path.isfile(in_path):
        print(f"Fichier introuvable: {in_path}", file=sys.stderr)
        return 1

    ts = datetime.now(timezone.utc).strftime("%Y%m%d_%H%M%S")
    out_path = args.out or os.path.join(
        os.path.dirname(in_path), f"rag_eval_results_{ts}.csv"
    )
    out_path = os.path.abspath(out_path)

    import json

    rows_out = []
    with open(in_path, newline="", encoding="utf-8-sig") as f:
        reader = csv.DictReader(f)
        fieldnames = list(reader.fieldnames or [])
        extra = [
            "http_status",
            "latency_ms",
            "reponse_bot",
            "erreur",
            "run_at_utc",
        ]
        out_fields = fieldnames + [c for c in extra if c not in fieldnames]

        for row in reader:
            q = (row.get("question") or "").strip()
            run_at = datetime.now(timezone.utc).isoformat()
            if not q:
                row = {**row, "http_status": "", "latency_ms": "", "reponse_bot": "", "erreur": "question vide", "run_at_utc": run_at}
                rows_out.append({k: row.get(k, "") for k in out_fields})
                continue

            body = json.dumps({"chatInput": q}, ensure_ascii=False).encode("utf-8")
            t0 = time.perf_counter()
            code, text = post_json(args.url, body, args.timeout)
            ms = int((time.perf_counter() - t0) * 1000)

            err = ""
            if code != 200:
                err = f"HTTP {code}" if code >= 0 else text[:500]

            row = {
                **row,
                "http_status": str(code),
                "latency_ms": str(ms),
                "reponse_bot": text if code == 200 else "",
                "erreur": err,
                "run_at_utc": run_at,
            }
            rows_out.append({k: row.get(k, "") for k in out_fields})
            time.sleep(max(0.0, args.delay))

    os.makedirs(os.path.dirname(out_path) or ".", exist_ok=True)
    with open(out_path, "w", newline="", encoding="utf-8") as f:
        w = csv.DictWriter(f, fieldnames=out_fields, extrasaction="ignore")
        w.writeheader()
        w.writerows(rows_out)

    print(f"OK — {len(rows_out)} lignes écrites dans {out_path}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/run_rag_eval.mjs
/**
 * Même rôle que run_rag_eval.py (Node 18+).
 * Usage: node scripts/run_rag_eval.mjs [--url http://...] [--csv tests/rag_eval_questions.csv] [--out fichier.csv]
 */
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

function parseArgs() {
  const a = process.argv.slice(2);
  const o = {
    url: process.env.RAG_EVAL_WEBHOOK_URL || 'http://127.0.0.1:5678/webhook/chat',
    csv: path.join(__dirname, '..', 'tests', 'rag_eval_questions.csv'),
    out: '',
    delay: Number(process.env.RAG_EVAL_DELAY_SEC || 0.5) * 1000,
    timeout: Number(process.env.RAG_EVAL_TIMEOUT_SEC || 180) * 1000,
  };
  for (let i = 0; i < a.length; i++) {
    if (a[i] === '--url') o.url = a[++i];
    else if (a[i] === '--csv') o.csv = a[++i];
    else if (a[i] === '--out') o.out = a[++i];
    else if (a[i] === '--delay') o.delay = Number(a[++i]) * 1000;
    else if (a[i] === '--timeout') o.timeout = Number(a[++i]) * 1000;
  }
  return o;
}

function parseCsv(text) {
  const lines = text.split(/\r?\n/).filter((l) => l.trim());
  if (!lines.length) return { headers: [], rows: [] };
  const headers = [];
  let cur = '';
  let q = false;
  const first = lines[0];
  for (let i = 0; i < first.length; i++) {
    const c = first[i];
    if (c === '"') q = !q;
    else if ((c === ',' && !q) || i === first.length - 1) {
      if (i === first.length - 1 && c !== ',') cur += c;
      headers.push(cur.trim().replace(/^"|"$/g, '').replace(/""/g, '"'));
      cur = '';
    } else cur += c;
  }
  const rows = [];
  for (let L = 1; L < lines.length; L++) {
    const vals = [];
    cur = '';
    q = false;
    const line = lines[L];
    for (let i = 0; i < line.length; i++) {
      const c = line[i];
      if (c === '"') q = !q;
      else if (c === ',' && !q) {
        vals.push(cur.replace(/^"|"$/g, '').replace(/""/g, '"'));
        cur = '';
      } else cur += c;
    }
    vals.push(cur.replace(/^"|"$/g, '').replace(/""/g, '"'));
    const row = {};
    headers.forEach((h, j) => {
      row[h] = vals[j] ?? '';
    });
    rows.push(row);
  }
  return { headers, rows };
}

function escCsv(v) {
  const s = String(v ?? '');
  if (/[",\n\r]/.test(s)) return `"${s.replace(/"/g, '""')}"`;
  return s;
}

async function main() {
  const opts = parseArgs();
  const inPath = path.resolve(opts.csv);
  if (!fs.existsSync(inPath)) {
    console.error('Fichier introuvable:', inPath);
    process.exit(1);
  }
  const raw = fs.readFileSync(inPath, 'utf8');
  const { headers, rows } = parseCsv(raw);
  const extra = ['http_status', 'latency_ms', 'reponse_bot', 'erreur', 'run_at_utc'];
  const outFields = [...headers, ...extra.filter((e) => !headers.includes(e))];

  const ts = new Date().toISOString().replace(/[:.]/g, '-').slice(0, 19);
  const outPath = opts.out
    ? path.resolve(opts.out)
    : path.join(path.dirname(inPath), `rag_eval_results_${ts}.csv`);

  const outLines = [outFields.map(escCsv).join(',')];

  for (const row of rows) {
    const q = (row.question || '').trim();
    const runAt = new Date().toISOString();
    if (!q) {
      const r = { ...row, http_status: '', latency_ms: '', reponse_bot: '', erreur: 'question vide', run_at_utc: runAt };
      outLines.push(outFields.map((h) => escCsv(r[h] ?? '')).join(','));
      continue;
    }

    const ac = new AbortController();
    const tid = setTimeout(() => ac.abort(), opts.timeout);
    const t0 = Date.now();
    let code = -1;
    let body = '';
    let err = '';
    try {
      const res = await fetch(opts.url, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ chatInput: q }),
        signal: ac.signal,
      });
      code = res.status;
      body = await res.text();
      if (!res.ok) err = `HTTP ${code}`;
    } catch (e) {
      err = String(e.message || e);
    }
    clearTimeout(tid);
    const ms = Date.now() - t0;

    const r = {
      ...row,
      http_status: String(code),
      latency_ms: String(ms),
      reponse_bot: code === 200 ? body : '',
      erreur: code === 200 ? '' : err,
      run_at_utc: runAt,
    };
    outLines.push(outFields.map((h) => escCsv(r[h] ?? '')).join(','));

    await new Promise((r) => setTimeout(r, Math.max(0, opts.delay)));
  }

  fs.mkdirSync(path.dirname(outPath), { recursive: true });
  fs.writeFileSync(outPath, outLines.join('\n'), 'utf8');
  console.log(`OK — ${rows.length} lignes → ${outPath}`);
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
